pkg/diff/diff.go

```diff
@@ -187,58 +256,56 @@ func (t *TableDiff) CheckTableData(ctx context.Context) (bool, error) {
 
 // EqualTableData checks data is equal or not.
 func (t *TableDiff) EqualTableData(ctx context.Context) (equal bool, err error) {
-	var allJobs []*CheckJob
-
 	table := t.TargetTable
-	useTiDB := false
 
+	useTiDB := false
 	if t.TiDBStatsSource != nil {
 		table = t.TiDBStatsSource
 		useTiDB = true
 	}
 
-	allJobs, err = GenerateCheckJob(table, t.Fields, t.Range, t.ChunkSize, t.Collation, useTiDB)
-
+	chunks, fromCheckpoint, err := GetChunks(ctx, table, t.Fields, t.Range, t.ChunkSize, t.Collation, useTiDB)
 	if err != nil {
 		return false, errors.Trace(err)
 	}
-
-	checkNums := len(allJobs) * t.Sample / 100
-	checkNumArr := getRandomN(len(allJobs), checkNums)
-	log.Info("check jobs", zap.Int("total job num", len(allJobs)), zap.Int("check job num", len(checkNumArr)))
-	if checkNums == 0 {
+	if len(chunks) == 0 {
+		log.Warn("get 0 chunks, table is not checked", zap.String("table", dbutil.TableName(t.TargetTable.Schema, t.TargetTable.Table)))
 		return true, nil
 	}
 
 	checkResultCh := make(chan bool, t.CheckThreadCount)
 	defer close(checkResultCh)
 
+	checkWorkerCh := make([]chan *ChunkRange, 0, t.CheckThreadCount)
 	for i := 0; i < t.CheckThreadCount; i++ {
-		checkJobs := make([]*CheckJob, 0, len(checkNumArr))
-		for j := len(checkNumArr) * i / t.CheckThreadCount; j < len(checkNumArr)*(i+1)/t.CheckThreadCount && j < len(checkNumArr); j++ {
-			checkJobs = append(checkJobs, allJobs[checkNumArr[j]])
-		}
-		go func(checkJobs []*CheckJob) {
-			eq, err := t.checkChunkDataEqual(ctx, checkJobs)
-			if err != nil {
-				log.Error("check chunk data equal failed", zap.Error(err))
-			}
-			checkResultCh <- eq
-		}(checkJobs)
+		checkWorkerCh = append(checkWorkerCh, make(chan *ChunkRange, 10))
+		go t.checkChunksDataEqual(ctx, t.Sample < 100 && !fromCheckpoint, checkWorkerCh[i], checkResultCh)
 	}
 
-	num := 0
+	defer func() {
+		for _, ch := range checkWorkerCh {
+			close(ch)
+		}
+	}()
+
+	var checkedNum int
+	go func() {
+		for _, chunk := range chunks {
+			checkWorkerCh[chunk.ID%t.CheckThreadCount] <- chunk
```

Comment:
it may be blocked if current `checkWorkerCh[chunk.ID%t.CheckThreadCount]` chan is too slow even other channels are empty?

Reply:
just make the code simple